### PR TITLE
[TT-11032] Godoc CI lint action, config and apidef/oas

### PR DESCRIPTION
**Pull request by titpetric**
### **User description**
We generate outputs based on godocs from two different packages:

- `/config` that provides configuration for the gateway
- `/apidef/oas` that documents the OAS schema

Two lint commands have been added into the taskfile, and `apidef-oas:lint` is enforced:

```
task apidef-oas:lint    # reports 66 violations, out of those 56 undocumented fields
task config:lint            # reports 411 violations, out of those 74 undocumented fields
```

As we provide user facing documentation, it should conform to our documentation standards. This PR should serve as the template for correcting godocs over these areas, as the CI change ensures we don't regress over time.

This aims to fix the apidef/oas state with missing documentation.

___

### **PR Type**
Enhancement, Tests

___

### **Description**
- Added new includes in the main Taskfile for `config` and `apidef-oas` taskfiles.
- Updated the `lint` task in the main Taskfile to include `config:lint` and `apidef-oas:lint`.
- Created a new Taskfile for `apidef/oas` with tasks for running tests, linting, and updating dependencies.
- Created a new Taskfile for `config` with tasks for running tests, linting, and updating dependencies.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>Taskfile.yml</strong><dd><code>Update Taskfile to include config and OAS schema linting</code>&nbsp; </dd></summary>
<hr>

Taskfile.yml

<li>Added new includes for <code>config</code> and <code>apidef-oas</code> taskfiles.<br> <li> Updated <code>lint</code> task to include <code>config:lint</code> and <code>apidef-oas:lint</code>.<br>

</details>

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk/pull/6386/files#diff-cd2d359855d0301ce190f1ec3b4c572ea690c83747f6df61c9340720e3d2425e">+8/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>Taskfile.yml</strong><dd><code>Add Taskfile for OAS schema tasks</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

apidef/oas/Taskfile.yml

<li>Created new Taskfile for <code>apidef/oas</code>.<br> <li> Added tasks for running tests, linting, and updating dependencies.<br>

</details>

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk/pull/6386/files#diff-077a7173655cb910898c8cb17ad0be73ecb04086fa2644afa609dfa5eb0a2b4f">+33/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>Taskfile.yml</strong><dd><code>Add Taskfile for config tasks</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

config/Taskfile.yml

<li>Created new Taskfile for <code>config</code>.<br> <li> Added tasks for running tests, linting, and updating dependencies.<br>

</details>

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk/pull/6386/files#diff-06a53f68fb9fa3abd7e994f2a99c28d457789d14f42a29673f9968cba4a265a0">+33/-0</a>&nbsp; &nbsp; </td>

</tr>                    
</table></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

### CODE REVIEW

File: apidef/oas/middleware.go [hunk 1]
```diff
@@ -1416,6 +1425,7 @@ func (v *VirtualEndpoint) ExtractTo(meta *apidef.VirtualMeta) {
 	}
 }
 
+// EndpointPostPlugins is a list of EndpointPostPlugins. It's used where multiple plugins can be run.
```
Comment (jeffy-mathew) [inline apidef/oas/middleware.go:1428]: i don't think we can run multiple post plugins now, it's kept as an array for future purposes I assume

File: apidef/oas/Taskfile.yml [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+      - go fmt ./...
+      - go test -count=1 ./...
+
+  # lint target is run from CI
+  lint:
+    desc: "Ensure linter pass"
+    cmds:
+      - schema-gen extract -o - | schema-gen lint --summary -i -
+
+  deps:
```
Comment (dcs3spp) [inline apidef/oas/Taskfile.yml:23]: what is upgrade policy for schema-gen in oas and config Taskfile? Should it be manual or automated by including deps as part of lint task?

File: apidef/oas/authentication.go [hunk 3]
```diff
@@ -336,14 +336,32 @@ func (as *AuthSource) ExtractTo(enabled *bool, name *string) {
...
-	Algorithm        string     `bson:"algorithm,omitempty" json:"algorithm,omitempty"`
-	Header           string     `bson:"header,omitempty" json:"header,omitempty"`
-	Query            AuthSource `bson:"query,omitempty" json:"query,omitempty"`
-	Secret           string     `bson:"secret,omitempty" json:"secret,omitempty"`
-	AllowedClockSkew int64      `bson:"allowedClockSkew,omitempty" json:"allowedClockSkew,omitempty"`
-	ErrorCode        int        `bson:"errorCode,omitempty" json:"errorCode,omitempty"`
-	ErrorMessage     string     `bson:"errorMessage,omitempty" json:"errorMessage,omitempty"`
+	// Enabled activates signature validation.
+	// Tyk classic API definition: `auth_configs[X].validate_signature`.
+	Enabled bool `bson:"enabled" json:"enabled"` // required
```
Comment (dcs3spp) [inline apidef/oas/authentication.go:341]: should required status be included in the description

File: apidef/oas/authentication.go [hunk 4]
```diff
@@ -336,14 +336,32 @@ func (as *AuthSource) ExtractTo(enabled *bool, name *string) {
...
+	// Algorithm is the signature method to use.
+	// Tyk classic API definition: `auth_configs[X].signature.algorithm`.
+	Algorithm string `bson:"algorithm,omitempty" json:"algorithm,omitempty"`
+	// Header is the name of the header to consume.
+	// Tyk classic API definition: `auth_configs[X].signature.header`.
```
Comment (dcs3spp) [inline apidef/oas/authentication.go:346]: Should use Tyk Classic instead of Tyk classic?

File: apidef/oas/security.go [hunk 5]
```diff
@@ -95,19 +95,48 @@ func (s *OAS) extractTokenTo(api *apidef.APIDefinition, name string) {
...
+
+	// SkipKid controls skipping using the `kid` claim from a JWT (default behaviour).
+	// When this is true, the field configured in IdentityBaseField is checked first.
+	SkipKid bool `bson:"skipKid,omitempty" json:"skipKid,omitempty"`
+
+	// PolicyFieldName is a configurable claim name from which a policy ID is extracted.
+	// The policy is applied to the session as a base policy.
+	PolicyFieldName string `bson:"policyFieldName,omitempty" json:"policyFieldName,omitempty"`
+
+	// Deprecated: ClientBaseField is unused.
```
Comment (dcs3spp) [inline apidef/oas/security.go:122]: What should be used instead of ClientBaseField?

File: apidef/oas/security.go [hunk 6]
```diff
@@ -323,12 +353,23 @@ func (e *ExtractCredentialsFromBody) ExtractTo(api *apidef.APIDefinition) {
 
 // OAuth configures the OAuth middleware.
 type OAuth struct {
-	Enabled               bool `bson:"enabled" json:"enabled"` // required
-	AuthSources           `bson:",inline" json:",inline"`
+	// Enabled activates the OAuth middleware.
+	Enabled bool `bson:"enabled" json:"enabled"` // required
```
Comment (dcs3spp) [inline apidef/oas/security.go:357]: Should add required state in the description?

File: apidef/oas/security.go [hunk 7]
```diff
@@ -515,10 +563,16 @@ func (c *IntrospectionCache) ExtractTo(cache *apidef.IntrospectionCache) {
 	cache.Timeout = c.Timeout
 }
 
+// ExternalOAuth holds configuration for an external OAuth provider.
 type ExternalOAuth struct {
-	Enabled     bool `bson:"enabled" json:"enabled"` // required
+	// Enabled activates external oauth functionality.
```
Comment (dcs3spp) [inline apidef/oas/security.go:568]: Should add required status to description in this field and in fields below where required comment appears?

**Review by dcs3spp — CHANGES_REQUESTED** [review]
Thanks @titpetric for adding in the automation for linting godoc comments.

Reviewed and made minor suggestions relating to adding required status to field descriptions included in the PR.

File: apidef/oas/security.go (hunk 5 above)
Comment (titpetric) [reply apidef/oas/security.go:122]: Nothing, can't say what it was intended for

File: apidef/oas/middleware.go (hunk 1 above)
Comment (titpetric) [reply apidef/oas/middleware.go:1428]: Future, tests, anyway it's reasonable too keep it around (even if it's orphaned right now)

File: apidef/oas/authentication.go (hunk 4 above)
Comment (titpetric) [reply apidef/oas/authentication.go:346]: No, unless we want to change the convention in the future, follows what exists

File: apidef/oas/Taskfile.yml (hunk 2 above)
Comment (titpetric) [reply apidef/oas/Taskfile.yml:23]: CI:

- It's automated by CI (on every tooling change, weekly).
- https://github.com/TykTechnologies/tyk-github-actions/blob/main/docker/tools/latest/Dockerfile

Development:

For manual updates of tooling we don't have a policy, but if you rely on it then you should keep your local versions up to date. Docker is also an option, something like this should use the tooling that CI uses:

```
docker run --platform linux/amd64 --rm -v $PWD:$PWD -w $PWD tykio/ci-tools:latest task
```

**Review by dcs3spp — APPROVED** [review]

**Pull request merged**